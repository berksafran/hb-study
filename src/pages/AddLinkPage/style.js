import styled from "styled-components";

const AddLinkPageContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto;
  width: 50%;
  padding: 10px;

  @media only screen and (max-width: 850px) {
    width: 100%;
  }

  .title {
    font-weight: 800;
    padding: 10px 0 10px 0;
  }

  p {
    font-size: 1.1rem;
  }

  div {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    padding: 10px;
  }

  .returnHome {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    padding: 20px 20px 20px 0px;
    margin-top: 20px;
    z-index: 8000;

    .icon {
      font-size: 0.9em;
      margin-right: 10px;
    }
  }

  a {
    color: black;
    font-weight: bold;
    font-size: 0.9rem;

    :hover,
    :activeÂ  {
      color: black;
    }
  }

  .addLinkForm {
    width: 100%;
  }
  
  .inputText {
    padding: 0 !important;
    margin: 0 0 .2em 0;
    width: 100%;

    :hover,
    :active,
    :focus {
      border-color: black;
    }
  }

  .inputLabel {
    font-size: 1.1rem;
    margin: 0 0 .2em 0;
  }

  .feedbackLabel {
    margin: 0;
    padding: 0;
  }

  .submitButton {
    font-size: 1.4em;
    font-weight: 800;
    background-color: black;
    color: white;
    border-color: black;
    border-radius: 2em;
    padding: 0.2em 1.8em 0.2em 1.8em;

    :hover {
      background-color: white;
      color: black;
      border-color: white;
    }
  }
`;

export { AddLinkPageContainer };
