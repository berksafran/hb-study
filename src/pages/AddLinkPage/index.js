import React, { useState } from "react";
import { Row, Form, Button } from "react-bootstrap";
import { Link } from "react-router-dom";
import { FaArrowLeft } from "react-icons/fa";
import { AddLinkPageContainer } from "./style";
import _ from "lodash";
import ToastWarn from "../../components/ToastWarn";

function AddLinkPage() {
  const [validated, setValidated] = useState(false);
  const [linkName, setLinkName] = useState();
  const [linkURL, setLinkURL] = useState();
  const [showToast, setShowToast] = useState({
    status: false,
    name: ''
  });

  // This variable is for using Ref and clear all inputs of add link form.
  let addLinkForm = '';

  const handleSubmit = event => {
    const form = event.currentTarget;

    // Check input forms
    if (form.checkValidity() === false) {
      event.preventDefault();
      event.stopPropagation();
      setValidated(true);
      event.preventDefault();
    } else {
      setValidated(true);
      event.preventDefault();

      // Creating new link 
      var currentTime = new Date().getTime()
      const linksArray = JSON.parse(localStorage.getItem("linksArray"));
      const newLinkItem = {
        id: currentTime,
        linkName: linkName,
        linkURL: linkURL,
        modifiedDate: currentTime,
        voteCounter: 0,
        points: 0
      };

      // Is there any data in Localstorage? If yes, add new item to array.
      if (_.isArray(linksArray)) {
        let newLinksArray = [...linksArray, newLinkItem];
        localStorage.setItem("linksArray", JSON.stringify(newLinksArray));
      } else {
        // If not, create new array.
        let newLinksArray = [newLinkItem];
        localStorage.setItem("linksArray", JSON.stringify(newLinksArray));
      }
      
      setShowToast({
        status: true,
        name: linkName
      });

      addLinkForm.reset();
      setValidated(false);
    }
  };

   // Open or Close Toast
   function handleToast() {
    setShowToast({
      ...showToast,
      status: !showToast,
      name: linkName
    });
  }

  return (
    <AddLinkPageContainer>
    <ToastWarn showToast={showToast} handleToast={() => handleToast()} toastType="add" />
      <Row className="returnHome" id="returnHome">
        <FaArrowLeft className="icon" />
        <Link to="/">Return to List</Link>
      </Row>
      <Row className="mt-0">
        <h2 className="title">Add New Link</h2>
      </Row>
      <Row className="mt-0">
        <Form
          noValidate
          validated={validated}
          onSubmit={handleSubmit}
          className="addLinkForm"
          ref={ref => addLinkForm = ref}
        >
          <Form.Label className="inputLabel">Link Name:</Form.Label>
          <Form.Row>
            <Form.Group md="4" controlId="linkName" className="inputText">
              <Form.Control
                required
                onChange={e => setLinkName(e.target.value)}
                type="text"
                placeholder="e.g. Alphabet"
              />
            </Form.Group>
          </Form.Row>
          <Form.Label className="inputLabel">Link URL:</Form.Label>
          <Form.Row>
            <Form.Group md="4" controlId="linkURL" className="inputText">
              <Form.Control
                required
                type="url"
                placeholder="e.g. http://abc.xyz"
                onChange={e => setLinkURL(e.target.value)}
                ref={ref => {ref = linkURL}}
              />
            </Form.Group>
          </Form.Row>
          <Form.Row className="d-flex justify-content-end">
            <Button id="submitButton" className="btn btn-secondary submitButton" type="submit">
              ADD
            </Button>
          </Form.Row>
        </Form>
      </Row>
    </AddLinkPageContainer>
  );
}

export default AddLinkPage;
