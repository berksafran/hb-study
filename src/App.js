import React from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import { Jumbotron, Container } from "react-bootstrap";
import Header from "./components/Header";
import ListLinkPage from "./pages/ListLinkPage";
import AddLinkPage from "./pages/AddLinkPage";
import ErrorPage from "./pages/ErrorPage";

function App() {
  return (
    <>
      <Router>
        <Container lg="true">
          <Jumbotron style={{ backgroundColor: "white" }}>
            <Header />
            <Switch>
              <Route path="/" exact component={() => <ListLinkPage />} />
              <Route path="/addlink" exact component={() => <AddLinkPage />} />
              <Route component={() => <ErrorPage />} />
            </Switch>
          </Jumbotron>
        </Container>
      </Router>
    </>
  );
}

export default App;
