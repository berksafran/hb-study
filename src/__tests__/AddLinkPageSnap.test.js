import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import App from "../App";
import AddLinkPage from "../pages/AddLinkPage";
import { BrowserRouter as Router } from "react-router-dom";

describe("Components in AddLinkPage test", () => {
  test("Header component Snapshot", () => {
    const header_tree = renderer
      .create(
        <Router>
          <AddLinkPage />
        </Router>
      )
      .toJSON();
    expect(header_tree).toMatchSnapshot();
  });
});
