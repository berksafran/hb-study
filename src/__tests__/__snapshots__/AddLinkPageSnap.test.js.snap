// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components in AddLinkPage test Header component Snapshot 1`] = `
<div
  className="sc-dnqmqq ksutkd"
>
  <div
    className="sc-htoDjs hKfUjC"
  >
    <div
      aria-atomic="true"
      aria-live="assertive"
      className="fade toast toast"
      role="alert"
    >
      <div
        className="toast-body"
      >
        <h2>
          <strong>
            
          </strong>
           added.
        </h2>
      </div>
    </div>
  </div>
  <div
    className="returnHome row"
    id="returnHome"
  >
    <svg
      className="icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
      />
    </svg>
    <a
      href="/"
      onClick={[Function]}
    >
      Return to List
    </a>
  </div>
  <div
    className="mt-0 row"
  >
    <h2
      className="title"
    >
      Add New Link
    </h2>
  </div>
  <div
    className="mt-0 row"
  >
    <form
      className="addLinkForm"
      noValidate={true}
      onSubmit={[Function]}
    >
      <label
        className="inputLabel form-label"
      >
        Link Name:
      </label>
      <div
        className="form-row"
      >
        <div
          className="inputText form-group"
          md="4"
        >
          <input
            className="form-control"
            id="linkName"
            onChange={[Function]}
            placeholder="e.g. Alphabet"
            required={true}
            type="text"
          />
        </div>
      </div>
      <label
        className="inputLabel form-label"
      >
        Link URL:
      </label>
      <div
        className="form-row"
      >
        <div
          className="inputText form-group"
          md="4"
        >
          <input
            className="form-control"
            id="linkURL"
            onChange={[Function]}
            placeholder="e.g. http://abc.xyz"
            required={true}
            type="url"
          />
        </div>
      </div>
      <div
        className="d-flex justify-content-end form-row"
      >
        <button
          className="btn btn-secondary submitButton btn btn-primary"
          disabled={false}
          id="submitButton"
          type="submit"
        >
          ADD
        </button>
      </div>
    </form>
  </div>
</div>
`;
